### DISCUSSION TITLE
Why doesn't the realtime server guarantee delivery?

### DISCUSSION THREAD

**Discussion by cloudcompute** [supabase/supabase#21093] [Questions]
`Taken from Readme..
The server does not guarantee that every message will be delivered to your clients so keep that in mind as you're using Realtime.`

What is the reason for this.. probably this one? The server is broadcasting the messages to the subscribers `only once` and if some subscriber is not available, the server is not tracking the unavailability.

**Comment by GaryAustin1** [comment]
They list this :
`Postgres database runs out of disk space due to Write-Ahead Logging (WAL) buildup, which can crash the database and prevent Realtime server from receiving and broadcasting changes. This can be mitigated in the Realtime RLS version of this server by setting the Postgres config max_slot_wal_keep_size to a reasonable size.
Realtime server can crash due to a larger replication lag than available memory, forcing the creation of a new replication slot and resetting replication to read from the latest WAL data.
When Realtime server falls too far behind for any reason, for example disconnecting from database as WAL continues to build up, then database can delete WAL segments the server still needs to read from, for example after reconnecting.`

And if you ever drop a connection due to internet issues, power down of app or tabs going into background there is no queue so any changes that happened are missed.  You would need to reconnect and refetch any potentially missed data.

**Reply by cloudcompute** [reply]
Thx for the reply.

Transactional Outbox pattern could be a better solution.

**Reply by GaryAustin1** [reply]
Not sure what that comment means.  You have a better solution you are going to look at?  Or you are suggesting a better solution to Supabase?  If so and you want them to see it, you might post in the realtime repository.

**Comment by cloudcompute** [comment]
I am 'suggesting' few other solutions:

a. This [project](https://github.com/fraktalio/fstore-sql?tab=readme-ov-file#run-postgres) uses Supabase. But it does not use the `realtime` library. It simply uses 2 extensions: pg_cron and pg_net as stated in the above link.

b. Transactional Outbox Pattern

![image](https://github.com/supabase/supabase/assets/6523462/2612e7a4-4d05-4e40-835d-211a78c25d23)

[Transactional Outbox PostgreSQL Example Code](https://github.com/Zehelein/pg-transactional-outbox)

I don't know whether this code can be used in production.

**Comment by Rubenburdin** [comment]
Realtime is "fire and forget" by design. If a client is disconnected when a change happens, that event is gone. There's no replay mechanism.
The reasons are architectural: Realtime reads from the Postgres WAL via a replication slot, broadcasts to connected clients, and moves on. It doesn't persist a queue of undelivered messages per subscriber because that would require tracking state for every client, which doesn't scale for their infrastructure.
Transactional outbox is indeed the robust solution if you need guaranteed delivery. Basic idea:

`create table outbox (
  id serial primary key,
  event_type text,
  payload jsonb,
  processed_at timestamptz,
  created_at timestamptz default now()
);

-- Your trigger writes to outbox atomically with the main change
create trigger order_created
  after insert on orders
  for each row
  execute function write_to_outbox();`


Then a separate worker polls the outbox, delivers events, and marks them processed. The event lives in Postgres until confirmed delivered.
pg_cron + pg_net combo works similarly. Schedule a job that queries for unprocessed events and POSTs them. If the POST fails, retry next run.
For client reconnection scenarios specifically, the practical pattern is: use Realtime for instant updates when connected, but on reconnect always fetch current state from the database. Don't rely on the event stream being complete.
Realtime is optimized for "good enough" live updates, not event sourcing. If you need the latter, outbox pattern or something like pg_notify with a persistent consumer is the way.
